File: sdk/java/src/main/java/io/rfb/sdk/internal/ZbrtCodec.java
package io.rfb.sdk.internal;

import io.rfb.sdk.DecodeError;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;

/**
 * Strict ZBRT v1 payload codecs (PROTOCOL.md §3.2): u32 BE length-prefixed
 * strings/bytes; every decode rejects truncation AND trailing bytes. INTERNAL.
 */
public final class ZbrtCodec {
    private ZbrtCodec() {
    }

    /** Cursor over a payload buffer with strict bounds checking. */
    static final class Reader {
        private final byte[] buf;
        private int pos;

        Reader(byte[] buf) {
            this.buf = buf;
        }

        boolean hasRemaining() {
            return pos < buf.length;
        }

        private void need(int n) {
            if (buf.length - pos < n) {
                throw new DecodeError("truncated payload");
            }
        }

        int u8() {
            need(1);
            return buf[pos++] & 0xFF;
        }

        boolean flag() {
            return u8() != 0;
        }

        int u32() {
            need(4);
            int v = ((buf[pos] & 0xFF) << 24) | ((buf[pos + 1] & 0xFF) << 16)
                    | ((buf[pos + 2] & 0xFF) << 8) | (buf[pos + 3] & 0xFF);
            pos += 4;
            return v;
        }

        int i32() {
            return u32();
        }

        byte[] bytes(int n) {
            need(n);
            byte[] out = new byte[n];
            System.arraycopy(buf, pos, out, 0, n);
            pos += n;
            return out;
        }

        byte[] prefixedBytes() {
            int n = u32();
            if (n < 0) {
                throw new DecodeError("truncated payload");
            }
            return bytes(n);
        }

        String text() {
            return new String(prefixedBytes(), StandardCharsets.UTF_8);
        }
    }

    static void putText(ByteArrayOutputStream out, String s) {
        byte[] b = s.getBytes(StandardCharsets.UTF_8);
        putU32(out, b.length);
        out.writeBytes(b);
    }

    static void putPrefixed(ByteArrayOutputStream out, byte[] b) {
        putU32(out, b.length);
        out.writeBytes(b);
    }

    /** Optional text: presence flag 0/1, then the length-prefixed UTF-8 text when present. */
    static void putFlagText(ByteArrayOutputStream out, String s) {
        if (s != null) {
            out.write(1);
            putText(out, s);
        } else {
            out.write(0);
        }
    }

    static void putU32(ByteArrayOutputStream out, int v) {
        out.write((v >>> 24) & 0xFF);
        out.write((v >>> 16) & 0xFF);
        out.write((v >>> 8) & 0xFF);
        out.write(v & 0xFF);
    }

    private static void checkTrailing(Reader r) {
        if (r.hasRemaining()) {
            throw new DecodeError("trailing payload");
        }
    }

    private static void checkCount(int count) {
        if (count > 255) {
            throw new DecodeError("too many capabilities");
        }
    }

    // ---- Hello / HelloAck ------------------------------------------------

    public static byte[] encodeHello(String client, java.util.List<String> capabilities) {
        return encodeHelloLike(client, capabilities);
    }

    public static byte[] encodeHelloAck(String server, java.util.List<String> capabilities) {
        return encodeHelloLike(server, capabilities);
    }

    private static byte[] encodeHelloLike(String name, java.util.List<String> capabilities) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        putText(out, name);
        checkCount(capabilities.size());
        out.write(capabilities.size());
        for (String cap : capabilities) {
            putText(out, cap);
        }
        return out.toByteArray();
    }

    public static byte[] encodeOutput(int stream, byte[] data) {
        ByteArrayOutputStream out = new ByteArrayOutputStream(8 + data.length);
        out.write(stream);
        putPrefixed(out, data);
        return out.toByteArray();
    }

    public static byte[] encodeError(long code, String message) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        putU32(out, (int) code);
        putText(out, message);
        return out.toByteArray();
    }

    public static HelloAck decodeHelloAck(byte[] payload) {
        Reader r = new Reader(payload);
        String server = r.text();
        int n = r.u8();
        java.util.List<String> capabilities = new java.util.ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            capabilities.add(r.text());
        }
        checkTrailing(r);
        return new HelloAck(server, capabilities);
    }

    // ---- Execute ---------------------------------------------------------

    public static byte[] encodeExecute(java.util.List<String> argv, String cwd,
                                       byte[] stdin, long timeoutMs) {
        if (argv.size() > 255) {
            throw new DecodeError("too many arguments");
        }
        if (timeoutMs < 0 || timeoutMs > 0xFFFFFFFFL) {
            throw new DecodeError("timeout_ms out of u32 range");
        }
        ByteArrayOutputStream out = new ByteArrayOutputStream(48 + stdin.length);
        out.write(argv.size());
        for (String arg : argv) {
            putText(out, arg);
        }
        putFlagText(out, cwd);
        putPrefixed(out, stdin);
        putU32(out, (int) timeoutMs);
        return out.toByteArray();
    }

    // ---- Output ----------------------------------------------------------

    public static Output decodeOutput(byte[] payload) {
        Reader r = new Reader(payload);
        int stream = r.u8();
        byte[] data = r.prefixedBytes();
        checkTrailing(r);
        return new Output(stream, data);
    }

    public static Execute decodeExecute(byte[] payload) {
        Reader r = new Reader(payload);
        int argc = r.u8();
        java.util.List<String> argv = new java.util.ArrayList<>(argc);
        for (int i = 0; i < argc; i++) {
            argv.add(r.text());
        }
        String cwd = r.flag() ? r.text() : null;
        byte[] stdin = r.prefixedBytes();
        long timeoutMs = r.u32() & 0xFFFFFFFFL;
        checkTrailing(r);
        return new Execute(argv, cwd, stdin, timeoutMs);
    }

    // ---- Exit ------------------------------------------------------------

    public static byte[] encodeExit(int code, Long signal) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        putU32(out, code);
        if (signal != null) {
            out.write(1);
            putU32(out, signal.intValue());
        } else {
            out.write(0);
        }
        return out.toByteArray();
    }

    public static Exit decodeExit(byte[] payload) {
        Reader r = new Reader(payload);
        int code = r.i32();
        Long signal = null;
        if (r.flag()) {
            signal = (long) r.u32() & 0xFFFFFFFFL;
        }
        checkTrailing(r);
        return new Exit(code, signal);
    }

    // ---- Cancel ----------------------------------------------------------

    public static byte[] encodeCancel(String reason, byte[] target16) {
        ByteArrayOutputStream out = new ByteArrayOutputStream(24);
        putFlagText(out, reason);
        if (target16 != null) {
            if (target16.length != 16) {
                throw new DecodeError("cancel target must be 16 bytes");
            }
            out.write(1);
            out.writeBytes(target16);
        } else {
            out.write(0);
        }
        return out.toByteArray();
    }

    public static Cancel decodeCancel(byte[] payload) {
        Reader r = new Reader(payload);
        String reason = r.flag() ? r.text() : null;
        byte[] target;
        if (!r.hasRemaining()) {
            target = null; // legacy payload without the target byte
        } else if (r.flag()) {
            target = r.bytes(16);
        } else {
            target = null;
        }
        checkTrailing(r);
        return new Cancel(reason, target);
    }

    // ---- Fs --------------------------------------------------------------

    public static byte[] encodeFs(int op, String path, byte[] data) {
        ByteArrayOutputStream out = new ByteArrayOutputStream(24 + data.length);
        out.write(op);
        putText(out, path);
        putPrefixed(out, data);
        return out.toByteArray();
    }

    public static Fs decodeFs(byte[] payload) {
        Reader r = new Reader(payload);
        int op = r.u8();
        String path = r.text();
        byte[] data = r.prefixedBytes();
        checkTrailing(r);
        return new Fs(op, path, data);
    }

    // ---- Health / HealthAck ----------------------------------------------

    public static byte[] encodeHealth(boolean healthy, String message) {
        ByteArrayOutputStream out = new ByteArrayOutputStream(16);
        out.write(healthy ? 1 : 0);
        putFlagText(out, message);
        return out.toByteArray();
    }

    public static Health decodeHealth(byte[] payload) {
        Reader r = new Reader(payload);
        boolean healthy = r.flag();
        String message = r.flag() ? r.text() : null;
        checkTrailing(r);
        return new Health(healthy, message);
    }

    // ---- Error -----------------------------------------------------------

    public static ZbrtErrorPayload decodeError(byte[] payload) {
        Reader r = new Reader(payload);
        long code = r.u32() & 0xFFFFFFFFL;
        String message = r.text();
        checkTrailing(r);
        return new ZbrtErrorPayload(code, message);
    }

    // ---- value holders -----------------------------------------------------

    /** HelloAck value: server name + negotiated capabilities. */
    public static final class HelloAck {
        public final String server;
        public final java.util.List<String> capabilities;

        public HelloAck(String server, java.util.List<String> capabilities) {
            this.server = server;
            this.capabilities = capabilities;
        }
    }

    public record Output(int stream, byte[] data) {
    }

    public record Execute(java.util.List<String> argv, String cwd, byte[] stdin, long timeoutMs) {
    }

    public record Exit(int code, Long signal) {
    }

    public record Cancel(String reason, byte[] target) {
    }

    public record Fs(int op, String path, byte[] data) {
    }

    public record Health(boolean healthy, String message) {
    }

    public record ZbrtErrorPayload(long code, String message) {
    }
}

File: sdk/nodejs/src/zbrt-codec.ts
/**
 * Strict ZBRT v1 payload codecs (PROTOCOL.md §3.2): u32 BE length-prefixed
 * UTF-8 text, byte arrays, optional-flag forms. Mirrors the Java ZbrtCodec.
 * INTERNAL.
 */
import { DecodeError } from './errors.js';

export const MAX_CAPABILITIES = 255;
export const MAX_ARGV = 255;
export const MAX_U32 = 0xffffffff;

class Reader {
  #buffer: Buffer;
  #pos = 0;

  constructor(buffer: Buffer) {
    this.#buffer = buffer;
  }

  hasRemaining(): boolean {
    return this.#pos < this.#buffer.length;
  }

  #need(n: number): void {
    if (this.#buffer.length - this.#pos < n) {
      throw new DecodeError('truncated payload');
    }
  }

  u8(): number {
    this.#need(1);
    // `#need` above guarantees at least one readable byte.
    return this.#buffer[this.#pos++]!;
  }

  flag(): boolean {
    return this.u8() !== 0;
  }

  u32(): number {
    this.#need(4);
    const v = this.#buffer.readUInt32BE(this.#pos);
    this.#pos += 4;
    return v;
  }

  i32(): number {
    return this.u32() | 0;
  }

  bytes(n: number): Buffer {
    this.#need(n);
    const out = this.#buffer.subarray(this.#pos, this.#pos + n);
    this.#pos += n;
    return out;
  }

  prefixedBytes(): Buffer {
    const n = this.u32();
    return this.bytes(n);
  }

  text(): string {
    return this.prefixedBytes().toString('utf8');
  }
}

function putU32(chunks: Buffer[], v: number): void {
  chunks.push(Buffer.from([(v >>> 24) & 0xff, (v >>> 16) & 0xff, (v >>> 8) & 0xff, v & 0xff]));
}

function putText(chunks: Buffer[], text: string): void {
  const b = Buffer.from(text, 'utf8');
  putU32(chunks, b.length);
  chunks.push(b);
}

function putPrefixed(chunks: Buffer[], bytes: Uint8Array): void {
  putU32(chunks, bytes.length);
  chunks.push(Buffer.from(bytes));
}

function putFlagText(chunks: Buffer[], text: string | null | undefined): void {
  if (text !== null && text !== undefined) {
    chunks.push(Buffer.from([1]));
    putText(chunks, text);
  } else {
    chunks.push(Buffer.from([0]));
  }
}

function checkTrailing(r: Reader): void {
  if (r.hasRemaining()) {
    throw new DecodeError('trailing payload');
  }
}

function checkCount(count: number): void {
  if (count > MAX_CAPABILITIES) {
    throw new DecodeError('too many capabilities');
  }
}

function concat(chunks: Buffer[]): Buffer {
  return Buffer.concat(chunks);
}

// ---- Hello / HelloAck ------------------------------------------------

export function encodeHello(client: string, capabilities: readonly string[]): Buffer {
  return encodeHelloLike(client, capabilities);
}

export function encodeHelloAck(server: string, capabilities: readonly string[]): Buffer {
  return encodeHelloLike(server, capabilities);
}

function encodeHelloLike(name: string, capabilities: readonly string[]): Buffer {
  checkCount(capabilities.length);
  const chunks: Buffer[] = [];
  putText(chunks, name);
  chunks.push(Buffer.from([capabilities.length]));
  for (const cap of capabilities) {
    putText(chunks, cap);
  }
  return concat(chunks);
}

export interface HelloAck {
  server: string;
  capabilities: string[];
}

export function decodeHelloAck(payload: Uint8Array): HelloAck {
  const r = new Reader(Buffer.from(payload));
  const server = r.text();
  const n = r.u8();
  const capabilities: string[] = [];
  for (let i = 0; i < n; i++) {
    capabilities.push(r.text());
  }
  checkTrailing(r);
  return { server, capabilities };
}

// ---- Execute -----------------------------------------------------------

export interface Execute {
  argv: string[];
  cwd: string | null;
  stdin: Buffer;
  timeoutMs: number;
}

export function encodeExecute(
  argv: readonly string[],
  cwd: string | null | undefined,
  stdin: Uint8Array,
  timeoutMs: number,
): Buffer {
  if (argv.length > MAX_ARGV) {
    throw new DecodeError('too many arguments');
  }
  if (timeoutMs < 0 || timeoutMs > MAX_U32) {
    throw new DecodeError('timeout_ms out of u32 range');
  }
  const chunks: Buffer[] = [Buffer.from([argv.length])];
  for (const arg of argv) {
    putText(chunks, arg);
  }
  putFlagText(chunks, cwd ?? null);
  putPrefixed(chunks, stdin);
  putU32(chunks, timeoutMs);
  return concat(chunks);
}

export function decodeExecute(payload: Uint8Array): Execute {
  const r = new Reader(Buffer.from(payload));
  const argc = r.u8();
  const argv: string[] = [];
  for (let i = 0; i < argc; i++) {
    argv.push(r.text());
  }
  const cwd = r.flag() ? r.text() : null;
  const stdin = r.prefixedBytes();
  const timeoutMs = r.u32();
  checkTrailing(r);
  return { argv, cwd, stdin, timeoutMs };
}

// ---- Output ------------------------------------------------------------

export interface Output {
  stream: number;
  data: Buffer;
}

export function encodeOutput(stream: number, data: Uint8Array): Buffer {
  const chunks: Buffer[] = [Buffer.from([stream])];
  putPrefixed(chunks, data);
  return concat(chunks);
}

export function decodeOutput(payload: Uint8Array): Output {
  const r = new Reader(Buffer.from(payload));
  const stream = r.u8();
  const data = r.prefixedBytes();
  checkTrailing(r);
  return { stream, data };
}

// ---- Exit ---------------------------------------------------------------

export interface Exit {
  code: number;
  signal: number | null;
}

export function encodeExit(code: number, signal: number | null | undefined): Buffer {
  const chunks: Buffer[] = [];
  putU32(chunks, code);
  if (signal !== null && signal !== undefined) {
    chunks.push(Buffer.from([1]));
    putU32(chunks, signal);
  } else {
    chunks.push(Buffer.from([0]));
  }
  return concat(chunks);
}

export function decodeExit(payload: Uint8Array): Exit {
  const r = new Reader(Buffer.from(payload));
  const code = r.i32();
  const signal = r.flag() ? r.u32() : null;
  checkTrailing(r);
  return { code, signal };
}

// ---- Cancel --------------------------------------------------------------

export interface Cancel {
  reason: string | null;
  target: Buffer | null;
}

export function encodeCancel(reason: string | null | undefined, target16: Uint8Array | null): Buffer {
  const chunks: Buffer[] = [];
  putFlagText(chunks, reason);
  if (target16 !== null && target16 !== undefined) {
    if (target16.length !== 16) {
      throw new DecodeError('cancel target must be 16 bytes');
    }
    chunks.push(Buffer.from([1]));
    chunks.push(Buffer.from(target16));
  } else {
    chunks.push(Buffer.from([0]));
  }
  return concat(chunks);
}

export function decodeCancel(payload: Uint8Array): Cancel {
  const r = new Reader(Buffer.from(payload));
  const reason = r.flag() ? r.text() : null;
  let target: Buffer | null = null;
  if (!r.hasRemaining()) {
    target = null; // legacy payload without the target byte
  } else if (r.flag()) {
    target = r.bytes(16);
  } else {
    target = null;
  }
  checkTrailing(r);
  return { reason, target };
}

// ---- Fs --------------------------------------------------------------------

export interface Fs {
  op: number;
  path: string;
  data: Buffer;
}

export function encodeFs(op: number, path: string, data: Uint8Array): Buffer {
  const chunks: Buffer[] = [Buffer.from([op])];
  putText(chunks, path);
  putPrefixed(chunks, data);
  return concat(chunks);
}

export function decodeFs(payload: Uint8Array): Fs {
  const r = new Reader(Buffer.from(payload));
  const op = r.u8();
  const path = r.text();
  const data = r.prefixedBytes();
  checkTrailing(r);
  return { op, path, data };
}

// ---- Health / HealthAck ----------------------------------------------

export interface Health {
  healthy: boolean;
  message: string | null;
}

export function encodeHealth(healthy: boolean, message: string | null | undefined): Buffer {
  const chunks: Buffer[] = [Buffer.from([healthy ? 1 : 0])];
  putFlagText(chunks, message);
  return concat(chunks);
}

export function decodeHealth(payload: Uint8Array): Health {
  const r = new Reader(Buffer.from(payload));
  const healthy = r.flag();
  const message = r.flag() ? r.text() : null;
  checkTrailing(r);
  return { healthy, message };
}

// ---- Error -------------------------------------------------------------

export interface ZbrtErrorPayload {
  code: number;
  message: string;
}

export function encodeError(code: number, message: string): Buffer {
  const chunks: Buffer[] = [];
  putU32(chunks, code);
  putText(chunks, message);
  return concat(chunks);
}

export function decodeError(payload: Uint8Array): ZbrtErrorPayload {
  const r = new Reader(Buffer.from(payload));
  const code = r.u32();
  const message = r.text();
  checkTrailing(r);
  return { code, message };
}
